File: mt-pricing-frontend/lib/api/endpoints/billing.ts
"use client";

import env from "@/lib/env";
import { createSupabaseBrowserClient } from "@/lib/supabase/client";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type InvoiceType =
  | "STANDARD"
  | "CREDIT_MEMO"
  | "DEBIT_MEMO"
  | "PROFORMA"
  | "INTERCOMPANY";

export type InvoiceStatus = "draft" | "posted" | "cancelled" | "reversed";

export type EInvoiceStatus =
  | "not_required"
  | "pending"
  | "compliant"
  | "rejected";

export interface InvoiceLineRead {
  id: string;
  invoice_id: string;
  product_sku: string;
  so_line_id: string | null;
  description: string | null;
  qty: string;
  unit_price: string;
  discount_pct: string;
  tax_rate: string;
  line_total: string | null;
  tax_amount: string | null;
}

export interface InvoiceRead {
  id: string;
  invoice_number: string;
  invoice_type: InvoiceType;
  delivery_id: string | null;
  so_id: string | null;
  customer_id: string;
  invoice_date: string;
  due_date: string | null;
  subtotal: string | null;
  tax_amount: string;
  total_amount: string | null;
  currency: string;
  status: InvoiceStatus;
  accounting_document_id: string | null;
  payment_terms: string;
  e_invoice_status: EInvoiceStatus;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  lines: InvoiceLineRead[];
}

export interface BillingKPIs {
  dso: string | null;
  cei: string | null;
  time_to_invoice_avg_hours: string | null;
  e_invoice_compliance_pct: string | null;
  overdue_invoice_count: number;
  overdue_amount: string;
}

export interface ARAgingBucket {
  customer_id: string;
  current: string;
  days_1_30: string;
  days_31_60: string;
  days_61_90: string;
  days_90_plus: string;
  total_outstanding: string;
}

export interface ARAgingReport {
  as_of_date: string;
  buckets: ARAgingBucket[];
}

export interface DunningItem {
  invoice_id: string;
  invoice_number: string;
  customer_id: string;
  due_date: string;
  days_overdue: number;
  dunning_level: number;
  total_amount: string;
}

// ---------------------------------------------------------------------------
// API client helpers
// ---------------------------------------------------------------------------

async function getAuthHeaders(): Promise<Record<string, string>> {
  const supabase = createSupabaseBrowserClient();
  const {
    data: { session },
  } = await supabase.auth.getSession();
  if (!session?.access_token) throw new Error("No auth session");
  return {
    "Content-Type": "application/json",
    Authorization: `Bearer ${session.access_token}`,
  };
}

const BASE = `${env.NEXT_PUBLIC_API_URL ?? ""}/api/v1/billing`;

async function get<T>(path: string): Promise<T> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${BASE}${path}`, { headers });
  if (!res.ok) throw new Error(`GET ${path} → ${res.status}`);
  return res.json() as Promise<T>;
}

async function post<T>(path: string, body?: unknown): Promise<T> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${BASE}${path}`, {
    method: "POST",
    headers,
    body: body !== undefined ? JSON.stringify(body) : undefined,
  });
  if (!res.ok) throw new Error(`POST ${path} → ${res.status}`);
  return res.json() as Promise<T>;
}

// ---------------------------------------------------------------------------
// billingApi
// ---------------------------------------------------------------------------

export const billingApi = {
  /** GET /billing/kpis */
  getKpis: () => get<BillingKPIs>("/kpis"),

  /** GET /billing/invoices */
  listInvoices: (params?: {
    customer_id?: string;
    status?: InvoiceStatus;
    limit?: number;
    offset?: number;
  }) => {
    const qs = new URLSearchParams();
    if (params?.customer_id) qs.set("customer_id", params.customer_id);
    if (params?.status) qs.set("status", params.status);
    if (params?.limit !== undefined) qs.set("limit", String(params.limit));
    if (params?.offset !== undefined) qs.set("offset", String(params.offset));
    const query = qs.toString() ? `?${qs.toString()}` : "";
    return get<InvoiceRead[]>(`/invoices${query}`);
  },

  /** GET /billing/invoices/{id} */
  getInvoice: (id: string) => get<InvoiceRead>(`/invoices/${id}`),

  /** POST /billing/invoices/{id}/post */
  postInvoice: (id: string) => post<InvoiceRead>(`/invoices/${id}/post`),

  /** POST /billing/invoices/{id}/cancel */
  cancelInvoice: (id: string) => post<InvoiceRead>(`/invoices/${id}/cancel`),

  /** GET /billing/ar-aging */
  getArAging: (as_of_date?: string) => {
    const qs = as_of_date ? `?as_of_date=${as_of_date}` : "";
    return get<ARAgingReport>(`/ar-aging${qs}`);
  },

  /** GET /billing/dunning */
  getDunning: (params?: { customer_id?: string; level?: number }) => {
    const qs = new URLSearchParams();
    if (params?.customer_id) qs.set("customer_id", params.customer_id);
    if (params?.level !== undefined) qs.set("level", String(params.level));
    const query = qs.toString() ? `?${qs.toString()}` : "";
    return get<DunningItem[]>(`/dunning${query}`);
  },
};

File: mt-pricing-backend/app/schemas/billing.py
"""Schemas Pydantic — Billing & Facturación (EP-ERP-05)."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Invoice
# ---------------------------------------------------------------------------

class InvoiceLineCreate(BaseModel):
    product_sku: str
    so_line_id: UUID | None = None
    description: str | None = None
    qty: Decimal
    unit_price: Decimal
    discount_pct: Decimal = Decimal("0")
    tax_rate: Decimal = Decimal("5")


class InvoiceLineRead(BaseModel):
    id: UUID
    invoice_id: UUID
    product_sku: str
    so_line_id: UUID | None
    description: str | None
    qty: Decimal
    unit_price: Decimal
    discount_pct: Decimal
    tax_rate: Decimal
    line_total: Decimal | None
    tax_amount: Decimal | None

    model_config = {"from_attributes": True}


class InvoiceCreate(BaseModel):
    invoice_number: str
    invoice_type: str = "STANDARD"
    delivery_id: UUID | None = None
    so_id: UUID | None = None
    customer_id: str
    invoice_date: date | None = None
    due_date: date | None = None
    currency: str = "AED"
    payment_terms: str = "NET30"
    lines: list[InvoiceLineCreate] = Field(default_factory=list)


class InvoicePatch(BaseModel):
    due_date: date | None = None
    payment_terms: str | None = None
    status: str | None = None
    e_invoice_status: str | None = None


class InvoiceRead(BaseModel):
    id: UUID
    invoice_number: str
    invoice_type: str
    delivery_id: UUID | None
    so_id: UUID | None
    customer_id: str
    invoice_date: date
    due_date: date | None
    subtotal: Decimal | None
    tax_amount: Decimal
    total_amount: Decimal | None
    currency: str
    status: str
    accounting_document_id: UUID | None
    payment_terms: str
    e_invoice_status: str
    created_by: UUID | None
    created_at: datetime
    updated_at: datetime
    lines: list[InvoiceLineRead] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class InvoiceChain(BaseModel):
    invoice: InvoiceRead
    so: dict | None = None
    delivery: dict | None = None


# ---------------------------------------------------------------------------
# Dunning
# ---------------------------------------------------------------------------

class DunningLevelRead(BaseModel):
    id: UUID
    level: int
    days_overdue: int
    action: str
    fee_amount: Decimal
    is_active: bool

    model_config = {"from_attributes": True}


class DunningHistoryRead(BaseModel):
    id: UUID
    invoice_id: UUID
    customer_id: str
    dunning_level: int
    sent_at: datetime
    notes: str | None

    model_config = {"from_attributes": True}


class DunningEscalateRequest(BaseModel):
    notes: str | None = None


# ---------------------------------------------------------------------------
# E-Invoice
# ---------------------------------------------------------------------------

class EInvoiceSubmitRequest(BaseModel):
    standard: str = "ZATCA_PHASE2"
    seller_name: str | None = None
    vat_number: str | None = None


class EInvoiceSubmissionRead(BaseModel):
    id: UUID
    invoice_id: UUID
    standard: str
    submission_ref: str | None
    submitted_at: datetime | None
    response_code: str | None
    response_message: str | None
    status: str
    xml_payload: str | None
    qr_code: str | None
    retry_count: int
    created_at: datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Payment Promises
# ---------------------------------------------------------------------------

class PaymentPromiseCreate(BaseModel):
    invoice_id: UUID
    customer_id: str
    promised_date: date
    promised_amount: Decimal | None = None
    notes: str | None = None


class PaymentPromisePatch(BaseModel):
    promised_date: date | None = None
    promised_amount: Decimal | None = None
    status: str | None = None
    notes: str | None = None


class PaymentPromiseRead(BaseModel):
    id: UUID
    invoice_id: UUID
    customer_id: str
    promised_date: date
    promised_amount: Decimal | None
    status: str
    notes: str | None
    created_by: UUID | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# AR Aging
# ---------------------------------------------------------------------------

class ARAgingBucket(BaseModel):
    customer_id: str
    current: Decimal = Decimal("0")
    days_1_30: Decimal = Decimal("0")
    days_31_60: Decimal = Decimal("0")
    days_61_90: Decimal = Decimal("0")
    days_90_plus: Decimal = Decimal("0")
    total_outstanding: Decimal = Decimal("0")


class ARAgingReport(BaseModel):
    as_of_date: date
    buckets: list[ARAgingBucket]


# ---------------------------------------------------------------------------
# KPIs
# ---------------------------------------------------------------------------

class BillingKPIs(BaseModel):
    dso: Decimal | None = None
    cei: Decimal | None = None
    time_to_invoice_avg_hours: Decimal | None = None
    e_invoice_compliance_pct: Decimal | None = None
    overdue_invoice_count: int = 0
    overdue_amount: Decimal = Decimal("0")
